Remove base type rooting for types in rooted assemblies

vitek-karas opened the pull request:
When rooting entire assemblies NativeAOT also roots all base types of all the types in such assembly. Regardless of which assembly the base type comes from. Historically this was necessary to make certain combinations of generic instantiations to work, but that's no longer the case. The compiler has better mechanism of tracking necessary things for generics now.

This rooting behavior brings in more code than necessary. This is specifically problematic when using the assembly rooting as a mechanism to test library trim/AOT compatibility. If the tested library depends on another library which has some incompatible code in it, but it's not used, ideally the compiler should remove the unused code and thus not see the incompatibilities. Rooting entire base types sometimes breaks that behavior and produces unnecessary warnings.

The product change is really just "don't root the base type", all of the rest of the change is tests. Modified existing test to validate more things around rooting behavior across assemblies. And then infrastructure changes to make it possible to use this test from NativeAOT.

This brings the behavior of NativeAOT and trimmer much closer with regard to assembly rooting.

Fixes https://github.com/dotnet/runtime/issues/92271.

vitek-karas committed:
Remove base type rooting for types in rooted assemblies

When rooting entire assemblies NativeAOT also roots all base types of all the types in such assembly. Regardless of which assembly the base type comes from. Historically this was necessary to make certain combinations of generic instantiations to work, but that's no longer the case. The compiler has better mechanism of tracking necessary things for generics now.

This rooting behavior brings in more code than necessary. This is specifically problematic when using the assembly rooting as a mechanism to test library trim/AOT compatibility. If the tested library depends on another library which has some incompatible code in it, but it's not used, ideally the compiler should remove the unused code and thus not see the incompatibilities. Rooting entire base types sometimes breaks that behavior and produces unnecessary warnings.

The product change is really just "don't root the base type", all of the rest of the change is tests. Modified existing test to validate more things around rooting behavior across assemblies. And then infrastructure changes to make it possible to use this test from NativeAOT.

This brings the behavior of NativeAOT and trimmer much closer with regard to assembly rooting.

MichalStrehovsky reviewed (approved):
Looks good but tests might need adjusting. Please also run nativeaot-outerloop before mergimg

vitek-karas committed:
Limit the impact only to rooted assemblies

Rooting from rd.xml is kept as before (rooting a type will also root its base types). Given that rd.xml is already very problematic and tricky and people rely on all kinds of "accidental" behavior, it's not worth the break - there's not much complexity in keeping this behavior there.

Update trimming tests which started to fail after the infra updates. Mostly disable the tests for NativeAOT as they don't make sense on AOT.

vitek-karas commented:
I changed it to only "Stop" the old behavior for rooted assemblies. rd.xml rooting will remain as before.
It doesn't feel worth to possible break people with this change for rd.xml.

Updated the trimming tests accordingly.

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

vitek-karas committed:
Fix reflection in PriorityQueue test

vitek-karas commented:
/azp run runtime-nativeaot-outerloop

vitek-karas committed:
Fix some of the outerloop test failures

vitek-karas committed:
Fix one more test

vitek-karas commented:
/azp run runtime-nativeaot-outerloop

vitek-karas commented:
@MichalStrehovsky could you please take another look - specifically at the test changes.

MichalStrehovsky commented on src/libraries/System.Security.Cryptography/tests/DSATests.cs:
```diff
@@ -171,15 +171,15 @@ protected override void Dispose(bool disposing)
             public override bool VerifySignature(byte[] rgbHash, byte[] rgbSignature) => _dsa.VerifySignature(rgbHash, rgbSignature);
             protected override byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm) =>
-                (byte[])_dsa.GetType().GetMethod(
+                (byte[])typeof(DSA).GetMethod(
```
There are multiple implementations of DSA/ECDSA so this is technically getting a different method, but since it's virtual it probably doesn't matter. Just wanted to point out for double checking.

vitek-karas replied:
Thanks - I was aware of that, but as you stated, the method is declared as virtual on the `DSA` class, so unless some derived class does `new` this will work (and `new` would be really weird for a protected virtual method). If this was in product I would be much more worried, but for tests I think this is perfectly OK.

MichalStrehovsky reviewed (approved):
Still looks good, thanks!

agocke commented:
/backport to release/8.0
